Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.h

```diff
@@ -51,168 +52,51 @@ enum class WiimoteGroup
-};
+enum class NunchukGroup;
+enum class ClassicGroup;
+enum class GuitarGroup;
+enum class DrumsGroup;
+enum class TurntableGroup;
 
-// Used for a dynamic swing or
-// shake.  This is used to pass
-// in data that defines the dynamic
-// action
-struct DynamicConfiguration
+template <typename T>
+void UpdateCalibrationDataChecksum(T& data, int cksum_bytes)
 {
-  double low_intensity;
-  int frames_needed_for_low_intensity;
+  constexpr u8 CALIBRATION_MAGIC_NUMBER = 0x55;
 
-  double med_intensity;
-  // Frames needed for med intensity can be calculated between high & low
-
-  double high_intensity;
-  int frames_needed_for_high_intensity;
-
-  int frames_to_execute;  // How many frames should we execute the action for?
-};
-
-struct ADPCMState
-{
-  s32 predictor, step;
-};
-
-struct ExtensionReg
-{
-  u8 unknown1[0x08];
+  static_assert(std::is_same<decltype(data[0]), u8&>::value, "Only sane for containers of u8!");
 
-  // address 0x08
-  u8 controller_data[0x06];
-  u8 unknown2[0x12];
+  auto cksum_start = std::end(data) - cksum_bytes;
 
-  // address 0x20
-  u8 calibration[0x10];
-  u8 unknown3[0x10];
+  // Checksum is a sum of the previous bytes plus a magic value (0x55).
+  // Extension calibration data has a 2nd checksum byte which is
+  // the magic value (0x55) added to the previous checksum byte.
+  u8 checksum = std::accumulate(std::begin(data), cksum_start, CALIBRATION_MAGIC_NUMBER);
 
-  // address 0x40
-  u8 encryption_key[0x10];
-  u8 unknown4[0xA0];
-
-  // address 0xF0
-  u8 encryption;
-  u8 unknown5[0x9];
-
-  // address 0xFA
-  u8 constant_id[6];
-};
-#pragma pack(pop)
-
-void UpdateCalibrationDataChecksum(std::array<u8, 0x10>& data);
-
-void EmulateShake(AccelData* accel, ControllerEmu::Buttons* buttons_group, double intensity,
-                  u8* shake_step);
-
-void EmulateDynamicShake(AccelData* accel, DynamicData& dynamic_data,
-                         ControllerEmu::Buttons* buttons_group, const DynamicConfiguration& config,
-                         u8* shake_step);
-
-void EmulateTilt(AccelData* accel, ControllerEmu::Tilt* tilt_group, bool sideways = false,
-                 bool upright = false);
-
-void EmulateSwing(AccelData* accel, ControllerEmu::Force* swing_group, double intensity,
-                  bool sideways = false, bool upright = false);
-
-void EmulateDynamicSwing(AccelData* accel, DynamicData& dynamic_data,
-                         ControllerEmu::Force* swing_group, const DynamicConfiguration& config,
-                         bool sideways = false, bool upright = false);
-
-enum
-{
-  ACCEL_ZERO_G = 0x80,
-  ACCEL_ONE_G = 0x9A,
-  ACCEL_RANGE = (ACCEL_ONE_G - ACCEL_ZERO_G),
-};
+  for (auto& i = cksum_start; i != std::end(data); ++i)
+  {
+    *i = checksum;
+    checksum += CALIBRATION_MAGIC_NUMBER;
+  }
+}
 
 class Wiimote : public ControllerEmu::EmulatedController
 {
   friend class WiimoteReal::Wiimote;
 
 public:
-  enum
+  enum : u8
+  {
+    ACCEL_ZERO_G = 0x80,
+    ACCEL_ONE_G = 0x9A,
```

Comment:
two conflicting definitions of ONE_G - 0x9A in Wiimote and 0xB3 in Nunchuck. Is that correct?

Reply:
That's fine. These seemed to be typical values on real hardware. Any value should work in theory.